cura/Machines/VariantNode.py

```diff
@@ -0,0 +1,159 @@
+            return  # There should not be any materials loaded for this printer.
+
+        # Find all the materials for this variant's name.
+        if not self.machine.has_machine_materials:  # Printer has no specific materials. Look for all fdmprinter materials.
+            materials = container_registry.findInstanceContainersMetadata(type = "material", definition = "fdmprinter")  # These are ONLY the base materials.
+        else:  # Printer has its own material profiles. Look for material profiles with this printer's definition.
+            all_materials = container_registry.findInstanceContainersMetadata(type = "material", definition = "fdmprinter")
+            printer_specific_materials = container_registry.findInstanceContainersMetadata(type = "material", definition = self.machine.container_id)
+            variant_specific_materials = container_registry.findInstanceContainersMetadata(type = "material", definition = self.machine.container_id, variant = self.variant_name)  # If empty_variant, this won't return anything.
+            materials_per_base_file = {material["base_file"]: material for material in all_materials}
+            materials_per_base_file.update({material["base_file"]: material for material in printer_specific_materials})  # Printer-specific profiles override global ones.
+            materials_per_base_file.update({material["base_file"]: material for material in variant_specific_materials})  # Variant-specific profiles override all of those.
+            materials = materials_per_base_file.values()
+
+        # Filter materials based on the exclude_materials property.
+        filtered_materials = [material for material in materials if material["id"] not in self.machine.exclude_materials]
+
+        for material in filtered_materials:
+            base_file = material["base_file"]
+            if base_file not in self.materials:
+                self.materials[base_file] = MaterialNode(material["id"], variant = self)
+                self.materials[base_file].materialChanged.connect(self.materialsChanged)
+        if not self.materials:
+            self.materials["empty_material"] = MaterialNode("empty_material", variant = self)
+
+    ##  Finds the preferred material for this printer with this nozzle in one of
+    #   the extruders.
+    #
+    #   If the preferred material is not available, an arbitrary material is
+    #   returned. If there is a configuration mistake (like a typo in the
+    #   preferred material) this returns a random available material. If there
+    #   are no available materials, this will return the empty material node.
+    #   \param approximate_diameter The desired approximate diameter of the
+    #   material.
+    #   \return The node for the preferred material, or any arbitrary material
+    #   if there is no match.
+    def preferredMaterial(self, approximate_diameter: int) -> MaterialNode:
```

Comment:
I'd name this `getPreferredMaterial()` because it's a function.

Reply:
I guess it's a style thing. I don't think that every function that returns something useful should be prefixed with `get`.